【OJT_自販機システム】UC02「商品を購入する」機能のテストコードを作成しました

wnakayama opened the pull request.

wnakayama committed:
FIX classDiag, CREATE sequence02

wnakayama committed:
FIX classDiag, CREATE sequence02

wnakayama committed:
FIX classDiag, sequence01-02. CREATE sequenceInit

wnakayama committed:
export to png

wnakayama committed:
FIX classDiag, sequence

wnakayama committed:
export to png

wnakayama committed:
FIX sequence02

wnakayama committed:
CREATE testcode for UC02

UC02「商品を購入する」機能で登場するクラスのテストコードを作成
購入可否チェック機能の追加にともないDepositExcessCheckerクラスを改名
Mainクラス, UserInterfaceクラスは結合テストで機能を確認するため削除

W-Nakayama-24 commented:
- テストコード https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/9/commits/a2f802a776638a2a756fecdb98053e1f3c3f0ef3

ご確認いただきたいテストコードのファイルは以下の4つです。
実施したい単体テストの内容は、テストコード内のコメントで補足説明を入れました。

- test/pkg/VendingMachineTest.java

- test/pkg/StorageTest.java

- test/pkg/ProductTest.java

- test/pkg/CheckerTest.java

[参考]
- 機能仕様書
https://htmlpreview.github.io/?https://github.com/W-Nakayama-24/VendingMachine_CUI/blob/master/UseCase/UseCase_Description.html

- クラス図
https://user-images.githubusercontent.com/54921025/69396624-36f27880-0d26-11ea-8bed-00cdf8f10017.png

- シーケンス図
https://user-images.githubusercontent.com/54921025/69304865-8b81ef00-0c65-11ea-8762-5f586b79561d.png

wnakayama committed:
ADD case of soda,mix,unknown

自動販売機クラスのget商品情報、get在庫数メソッドで不足していたテストケースを追加

wnakayama commented on test/pkg/VendingMachineTest.java:
```diff
@@ -8,53 +8,99 @@
+    public void getStockFailure() {
+        VendingMachine vm = new VendingMachine();;
+        vm.getStock(9);
+    }
```
VendingMachineTest.javaのget在庫数メソッドで一部記述漏れがありました。
確認が行き届いておらず申し訳ないですが、併せてご覧ください。

wnakayama commented on test/pkg/VendingMachineTest.java:
```diff
@@ -8,53 +8,99 @@
+    @Test(expected = WrongProductNumberException.class)
+    public void getProductFailure() {
+        VendingMachine vm = new VendingMachine();
+        vm.getProduct(9);
```
VendingMachineTest.javaのget商品情報メソッドで一部記述漏れがありました。
確認が行き届いておらず申し訳ないですが、併せてご覧ください。

yutsunom commented on test/pkg/VendingMachineTest.java:
```diff
@@ -8,31 +8,99 @@
+        assertThat(result,is(12));
+    }
+
+    //サイコソーダ、ミックスオレも同様に実施
```
テスト繰り返しは、配列、Array などで表現できると思います。
["おいしい水", "サイコソーダ", "ミックスオレ"]
を、for 文なり each(?) なりでまわす。

たとえば
```java
String[] products =  ["おいしい水", "サイコソーダ", "ミックスオレ"]
for (int i = 0; i < products.length; i++) {
        VendingMachine vm = new VendingMachine();
        Product result = vm.getProduct(1);
        assertThat(result.getNum(), is(1));
        assertThat(result.getName(), is(products[ i ]);
        assertThat(result.getPrice(), is(100));
}
```

yutsunom commented on test/pkg/VendingMachineTest.java:
```diff
@@ -8,31 +8,99 @@
 
     @Test
-    // getterのテスト 最初の投入金額0円を取得する
+    //getterのテスト 最初の投入金額0円を取得する
```
コメントは javadoc で書けばよいと思います。

yutsunom commented on test/pkg/ProductTest.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Test
+    public void testGetNum() {
+        Product prdct = new Product(1, null, 0);
```
変数名は省略しないほうが無難かと。分からなくなります。

yutsunom replied:
```suggestion
        Product product = new Product(1, null, 0);
```

yutsunom commented on test/pkg/StorageTest.java:
```diff
@@ -0,0 +1,130 @@
+    @Test
+    public void getWaterSuccess() {
+        Storage storage = new Storage();
+        Product testWater = storage.getProduct(1);
```
getProduct はget在庫数かな
クラス図のほう、名前更新してください

wnakayama replied:
getProduct が get商品情報メソッドで、
getStockが get在庫数メソッドに相当します。

yutsunom commented on test/pkg/StorageTest.java:
```diff
@@ -0,0 +1,130 @@
+
+import org.junit.Test;
+
+public class StorageTest {
```
Storage クラスには 商品補充するメソッドがあるはずで、
それがないと商品番号がハードコードになりますが、意図したところでしょうか。
それでいうと、ユースケースに「商品を補充する」がないとダメな気がする

yutsunom replied:
テストとしては、テストクラス内でたとえば
```java
Product deliciasWater = new Product(1, "おいしい水", 100); // 1 商品コード, 100 値段
for (int i =0; i<100; i++) {
  storage.addProduct(deliciasWater);
}
```
とかする必要があるかと。そうしないと、テストの `getProduct(1);` の１が美味しい水だという情報がハードコードになる

wnakayama replied:
商品補充のユースケースが無い代わりにStorageクラスでサンプルの3商品をセットします。

先ほど指摘した頂いたように、コンストラクタではなくメソッドを用意したうえでもう一度お見せします

yutsunom commented on test/pkg/StorageTest.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    // getProductメソッドのテスト 存在しない商品番号を指定した場合、不正なインデックスを使って配列にアクセスしてしまう
+    // 商品番号が正しくないことを伝えるため、自作例外WrongProductNumberExceptionが発生
```
このコメントって、レビュー用？だったら、github のコメント機能で書いてもらって、テストコードからは消しましょう

yutsunom commented on test/pkg/StorageTest.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    // getStockメソッドのテスト
+    // 商品番号を指定し、コンストラクタでstockMapにputしていた在庫数が正しく取得できるかどうかでテストする
```
コンストラクタでstockMap（Map）に直接putはしなくないですか？
商品が追加されるとき、中で更新するのだと思います。
クラス図に、各クラスコンストラクタの引数を書いてみてください

wnakayama replied:
現在は引数無しコンストラクタを使っていますが、修正で必要になった際にはクラス図にも反映させます。

yutsunom commented on test/pkg/StorageTest.java:
```diff
@@ -0,0 +1,130 @@
+    // getStockメソッドのテスト
+    // productInfoMapに存在しない商品番号(キー)を指定した場合
+    // 対応する番号を持った商品が見つからなかった結果、nullを返却してしまう
+    // 商品番号が正しくないことを伝えるため、自作例外WrongProductNumberExceptionが発生
```
商品番号は正しいが、商品が見つからない場合は０が返るとかですか？
最初の補充前はMap に該当する商品番号のキーが入ってない可能性は、あるのではないでしょうか。

それよりは、Product[] のフィールドをみて管理したほうがいいかと。（MTの話ではないですが）

yutsunom replied:
productInfoMap と stockMap は違うもの？？

wnakayama replied:
「商品情報(番号/名前/単価)」のテーブルと「在庫数」のテーブルを用意する、というデータベース的な考え方を使っています。

初めに呼び出す商品補充のメソッドを実装してみて、このあたりの修正が必要になるか検討します。

yutsunom reviewed (changes requested):
取り急ぎ（あとは明日）

yutsunom commented on test/pkg/VendingMachineTest.java:
```diff
@@ -41,12 +109,91 @@ public void insertFailure() {
 
-        boolean insertResult = vm.insertMoney(requestMoney);
-        assertThat(insertResult, is(false));
+        boolean insertResult  = vm.insertMoney(requestMoney);
```
requestMoney 定義してない

yutsunom commented on test/pkg/VendingMachineTest.java:
```diff
@@ -8,31 +8,99 @@
     @Test
-    // お金投入のテスト 希望金額9991円で上限チェックがNGのとき、falseを返す
-    // 投入金額が変化していない
+    //お金投入のテスト 希望金額9991円で上限チェックがNGのとき、falseを返す
```
9991円 って、どこで定義してますか？

wnakayama replied:
> requestMoney 定義してない

> 9991円 って、どこで定義してますか？

すみません、githubの方で定義部分(L107~108)の表示が折りたたまれていたようです
(冒頭数行のimport文なんかが折りたたまれるのは分かるんですが、所々でこのような折りたたみが行われているのは何故でしょうか…)

▼折り畳み前
![exp1](https://user-images.githubusercontent.com/54921025/69504563-4a405680-0f67-11ea-9ce0-0772d4bc4647.PNG)

▼折り畳み後
![exp2](https://user-images.githubusercontent.com/54921025/69504564-4a405680-0f67-11ea-8122-9ace3d4e8a9a.PNG)

yutsunom commented on test/pkg/VendingMachineTest.java:
```diff
@@ -41,12 +109,91 @@ public void insertFailure() {
+        int stockBeforeBuy = vm.getStock(1);
+        vm.deposit = depositBeforeBuy;
+
+        boolean result = vm.buyProduct(1);
```
真偽値というか、「投入金額が足りていない」ことを表すリザルトコードもしくはそういう例外を返したほうがいいです。
たとえば、客に「投入金額が足りてなかったので失敗した」という情報を伝えたいとき、どう実現するか考えてみましょう。（ログとか、画面とかに書くと思います）

wnakayama replied:
利用者へのエラー内容の伝達は、最終的に自販機クラスの呼び出し元であるUserInterfaceクラスが戻り値に応じてエラーメッセージを出力するという形をとっています。

今回エラーメッセージの標準出力はJUnitを使わず結合テストで見るようにしていまして、
その関係でUserInterfaceクラスのテストコードが無く、テストコードのプルリクだけではエラー内容を利用者に伝える術が用意されていないように見えていたかと思います。申し訳ないです…

![seq1125](https://user-images.githubusercontent.com/54921025/69506784-74981100-0f73-11ea-9e2d-6294617355f9.PNG)

yutsunom replied:
でしたら、 UserInterface クラスに、VenderMachine クラスの戻り値が必要かと思います。
今 alt を通ろうがその下を通ろうが、false しか返ってこないので、どんなエラーが起こったかを判断できなくないですか

yutsunom commented on test/pkg/VendingMachineTest.java:
```diff
@@ -41,12 +109,91 @@ public void insertFailure() {
+        int stockBeforeBuy = vm.getStock(1);
+        vm.deposit = depositBeforeBuy;
+
+        boolean result = vm.buyProduct(1);
```
同上。例外なりコードなり返しましょう

yutsunom commented on test/pkg/StorageTest.java:
```diff
@@ -0,0 +1,130 @@
+        int beforeTest = storage.getStock(1);
+        assertThat(beforeTest, is(12));
+
+        storage.reduceStock(1);
+
+        int afterTest = storage.getStock(1);
```
reduceStock の戻り値でストック減らした後の数字返すとかすればまとめられる気がします。
（現状のreduceStock の戻り値何でしたっけ）

W-Nakayama-24 replied:
現状reduceStockの戻り値はvoidで設計してありますが、確かに減らした後の数字を戻り値にすると良いですね。修正します。

wnakayama replied:
現状reduceStockの戻り値はvoidで設計してありますが、確かに減らした後の数字を戻り値にすると良いですね。修正します。

yutsunom commented on test/pkg/CheckerTest.java:
```diff
@@ -0,0 +1,108 @@
+public class CheckerTest {
+
+    @Test
+    // 希望金額が1000円,投入金額が0円のとき,加算後に上限の9990円を超えないのでtrueを返す
```
「希望金額」って何ですか・・？

yutsunom replied:
希望金額：投入金額（これから投入するつもりの金額）
投入金額：現在の残高
のことかな？

yutsunom replied:
そうだとすると、checkDepositExcess ってなんで定義したんでしょうか
せめて、残高がブラックボックスだとして、
checkDepositExcess (投入金額（これから投入するつもりの金額）)として、
 戻り値：投入金額あわせた残高

のように引数1個だったらまだわかるかな。。

yutsunom commented on test/pkg/CheckerTest.java:
```diff
@@ -0,0 +1,108 @@
+        Checker affchecker = new Checker();
+        Storage storage = new Storage();
+        Product product = storage.getProduct(1);
+        boolean checkResult = affchecker.checkCanAfford(product, 100);
```
checkCanAfford　わざわざ定義するかな
たとえばProduct のほうに、希望金額のgetter があれば、
product.getUnitPrice() < 100 とか1行で比べられるのでは

wnakayama replied:
>checkDepositExcess ってなんで定義したんでしょうか

>checkCanAfford　わざわざ定義するかな

オブジェクト指向やクラス設計のセオリーが未だ掴みきれておらず、
自販機クラスの仕事は少しでも別クラスに分けておいたほうが良いかと考えていました

どちらのチェックも試しにソースコードを書いてみると大した事をやっていないので
自販機クラスにそのままやらせても良かったと思います

Checkerクラスを統合するかどうかは、もう少し考えさせてください。

wnakayama committed:
CREATE class storage

W-Nakayama-24 commented:
補足で、現在試作しているStorageクラスのソースコードをpushしました。

https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/9/files

githubにはテストコードしかpushしていないのが原因でこちらの意図している処理が分かりづらくなっていたかと思います。申し訳ありません。

wnakayama commented on Source/VendingMachine/src/pkg/Storage.java:
```diff
@@ -0,0 +1,53 @@
+        stockMap.put(1, 12);
+        stockMap.put(2, 5);
+        stockMap.put(3, 1);
+    }
```
今回「商品を補充する」ユースケースが無く、データベースにサンプルデータを挿入するといった事もおこないませんが代替策としてStorageクラスのコンストラクタでこのように3つの商品を用意しています。

yutsunom replied:
(slack にコメントした内容)
Storage クラスに、商品追加する public メソッドを作って、
main 関数（Storage クラスを new する）から、呼び出しましょう。
`Product water = new Product(1, "おいしい水", 100);` などは、main 関数から呼ぶ。

そうすると、テストクラスでも `Product water = new Product(1, "おいしい水", 100);` できるようになる

wnakayama committed:
ADD javadoc, RENAME variable, FIX BuyProduct

すべてのテストクラスでコメントをjavadoc形式に変更
ProductTestクラスにてprdct→productに命名を変更
VendingMachineTestクラス - BuyProductメソッドの戻り値をenumに変更

W-Nakayama-24 commented:
途中ですが、以下のように修正を実施したので報告します。

- すべてのテストクラスでコメントをjavadoc形式に変更

- ProductTestクラスにてprdct→productに命名を変更

- VendingMachineTestクラス - BuyProductメソッドの戻り値をenumに変更
　それにともないアサーションにおける期待値を変更 

※ 設計時の戻り値(boolean)では最終的な出力メッセージを
　「投入金額不足」「指定商品の在庫切れ」「購入成功」の3種類に分けることが困難だったため。
　11/22(金)に質問・相談を実施し、列挙型(enum)でチェック結果を返すように変更

    enum BuyResult{
        SUCCESS,
        NOT_ENOUGH_MONEY,
        ZERO_STOCK
    }

wnakayama committed:
REMOVE constructor, ADD methods for init

wnakayama commented on Source/VendingMachine/src/pkg/Storage.java:
```diff
@@ -0,0 +1,53 @@
+        stockMap.put(1, 12);
+        stockMap.put(2, 5);
+        stockMap.put(3, 1);
+    }
```
>せめて、Storage クラスに 商品をpush するメソッドをpublic でつくって、その中で数増やしましょう

Storageクラスのコンストラクタ内で3つの商品と在庫を初期化していた部分を、2つのメソッドに分けました。

1点伺いたいのですが、これら2つのメソッドを実行する場所として
Storageのコンストラクタではなく
Storageクラスをnewする個所のほうがより適切である理由は何でしょうか?

現状Storageクラスを初めてnewしているのはVendingMachineクラスなので
VendingMachineクラスのコンストラクタで2つのメソッドを呼び出しているのですが

Storageクラスの単体テストで、VendingMachineクラスは登場しないため以下のコード(※)が新しく必要になり
「こうなるくらいならStorageクラスのコンストラクタでも良いのでは?」と思いました。

    @Test
    public void getWaterSuccess() {
        Storage storage = new Storage();
        storage.initProduct(); ※
        storage.initStorage(); ※
        Product testWater = storage.getProduct(1);
        assertThat(testWater.getNum(),is(1));
        assertThat(testWater.getName(),is("おいしい水"));
        assertThat(testWater.getPrice(),is(100));
    }

yutsunom replied:
（さきほど対面で話した内容）
・Storage クラスに public で chargeProduct(Product product) を作る
・VendingMachine クラスにも public で chargeProduct(Product product) を作る
・main 関数からは、VendingMachine に対して chargeProduct する
・VendingMachine クラスからは、Storage に対して chargeProduct する
・main 関数が Product を知るのは良いと思う。補充するとき商品情報知らないといけないので
・現状、main 関数は Storage クラスとは関連させないほうがよい。

wnakayama replied:
こちらの内容でソースコードを修正中です。

①現在、mainクラスとVendingMachineクラスの間にUserInterfaceクラスがあって、
"UserInterfaceクラスで"VendingMachineをnewしているという状況だったのですが

これもmainでnewするようにして、利用者向けのメッセージを出す仕事が主なUserInterfaceクラスの各メソッドには(VendingMachine vm)を引数に渡す形にしたほうが適切でしょうか?

②サンプル3商品の情報を格納する為にStorageクラスでProduct型の配列を長さ3で用意していますが
後から商品を追加することを考えると配列ではなく可変長のArrayListに変えるほうが適切でしょうか?

以上2点ご回答をお願いします。

yutsunom replied:
2 はYes。ですが、Map で管理するんでなかったでしたっけ
1 は UserInterface 仕様によりますが、たぶんUserInterface クラスでやるようになるのかな。
入出力を担当する箇所かな

wnakayama replied:
Mapで管理してあるのは在庫数の方ですね。
商品番号をキーにして、紐づいている在庫数を取ってくるという使い方をします

「例えばソーダのボトル1本に在庫という情報が果たして備わっているのか?」という観点から、商品情報は別のコレクションで管理する方法をとっていたので
配列か? ArrayListか?という質問をさせていただきました

ご回答ありがとうございます!

W-Nakayama-24 commented:
次の実施内容

- javadocコメントのうち、レビュー時にgithubで説明していれば十分な箇所は削る
- WrongProductNumberExceptionが発生しうるテストメソッドにthrows宣言を追加
　　- この自作例外がどんな条件で発生するのか、説明を追加

W-Nakayama-24 commented:
[口頭で頂いたレビューのメモ]

- 将来的に商品補充のユースケースを実現したいときのことを考えてみる
 VendingMachineクラスにChargeProductメソッドを追加することを考える(引数はint Str int)

- どういったProductを管理するかは、mainが知っていてもいいのではないか(productを外側に出す)
Storageを隠しておくのは良い
なぜ ⇒ ハードコード過ぎる 補充のユースケースが想定されていない為、本当にサンプルの3商品しか扱えない
拡張性を保つためにも、(ユースケースとして当初出せていなかったにしても)内部の仕組みとして補充がきくようにしたほうが良いのではないか

W-Nakayama-24 commented:
>①現在、mainクラスとVendingMachineクラスの間にUserInterfaceクラスがあって、
"UserInterfaceクラスで"VendingMachineをnewしているという状況だったのですが
これもmainでnewするようにして、利用者向けのメッセージを出す仕事が主なUserInterfaceクラスの各メソッドには(VendingMachine vm)を引数に渡す形にしたほうが適切でしょうか?

>>1 は UserInterface 仕様によりますが、たぶんUserInterface クラスでやるようになるのかな。

サンプルの3商品を、UserInterfaceクラスのコンストラクタで用意するようにソースコードを修正しました。



    UserInterface() {
        Product water = new Product(1,"おいしい水",100);
        Product soda = new Product(2,"サイコソーダ",150);
        Product mix = new Product(3,"ミックスオレ",160);
        vm.chargeProduct(water);
        vm.chargeProduct(soda);
        vm.chargeProduct(mix);

        vm.chargeStorage(1, 12);
        vm.chargeStorage(2, 5);
        vm.chargeStorage(3, 1);
    }

W-Nakayama-24 commented:
今回の修正で、テストクラスでサンプルとは異なる価格の商品や、在庫0の商品を新しく用意してみることが可能になりました。

逆に、これまでのサンプル3商品を使ったテストコードはそのままだと実行できなくなったのですが、
(UserInterfaceクラスで登場するようになった為、他クラスの単体テストではnew Productとして別途用意する必要がある)
　
テストクラス内でテストケースに応じた商品を補充できるようになった以上、サンプル3商品のハードコードなテストは削除しておいたほうが良いでしょうか？

yutsunom commented:
> 
> 
> 今回の修正で、テストクラスでサンプルとは異なる価格の商品や、在庫0の商品を新しく用意してみることが可能になりました。
> 
> 一方で、これまでのサンプル3商品を使ったテストコードはそのままだと実行できなくなったのですが、
> (UserInterfaceクラスで登場するようになった為、他クラスの単体テストではnew Productとして別途用意する必要がある)
> 　
> テストクラス内でテストケースに応じた商品を補充できるようになった以上、サンプル3商品のハードコードなテストは削除しておいたほうが良いでしょうか？

ハードコードなテストは必要です。
> 他クラスの単体テストではnew Productとして別途用意する必要がある
を実施できるようになったと思います。
・テストクラスでProduct を追加して、
・それを確認する
というのが今回できるようになりました。assertXX の直前に追加したパラメータと一致するので、確認が容易になったというか、明瞭にできるようになったということです

W-Nakayama-24 commented:
public class StorageTest {

    /**
     * chargeProductメソッドのテスト
     * UserInterfaceクラスで用意するサンプルとは異なる商品を追加する。
     * それをgetメソッドを使って正しく取得できていれば、chargeProductが成功していたと言える??
     * @throws WrongProductNumberException 商品番号として存在していない整数を引数に渡すと発生
     */
    @Test
    public void chargeProductSuccess() throws WrongProductNumberException {
        Storage storage = new Storage();
        Product productForTest = new Product(4,"テスト用商品",500);
        storage.chargeProduct(productForTest);

        Product productFromList = storage.getProduct(1);
        assertThat(productFromList.getNum(),is(4));
        assertThat(productFromList.getName(),is("テスト用商品"));
        assertThat(productFromList.getPrice(),is(500));
    }

wnakayama committed:
FIX storageTest, checkerTest

wnakayama committed:
FIX src_storage

wnakayama commented on test/pkg/StorageTest.java:
```diff
@@ -0,0 +1,168 @@
+        assertThat(waterFromList.getNum(), is(1));
+        assertThat(waterFromList.getName(), is("おいしい水"));
+        assertThat(waterFromList.getPrice(), is(100));
+    }
```
以下の内容を実施しました。

- chargeProductメソッドのテストコードを追加
- getProductメソッドのテストコードを修正
　　- テストクラス内でテスト対象のProductを生成し、それをアサーションで確認する方式に変更

wnakayama replied:
chargeProductメソッドですが、どこまでテストコードで確かめるべきでしょうか?
以下の2案で迷っています。

① UserInterfaceのコンストラクタと異なる商品を、テストクラス内で新しくListに追加する処理までを確認する(pushしたテストコードの通り)

②テストクラス内で新しくListに追加した商品を、getメソッドを使ってListから再び取得 ⇒ 取得結果(商品番号、商品名、単価)をアサーションで確認


②だと、getProductメソッドのテストと同じになってしまうと考えたのでpushしたテストコードは①の案を採用しています

wnakayama commented on Source/VendingMachine/src/pkg/Storage.java:
```diff
@@ -1,53 +1,38 @@
-
+        Product pr = productList.get(num - 1);
+        return pr;
     }
```
chargeProductメソッド、chargeStockメソッドを追加した後のStorageクラスのソースコードをpushしました。

「引数で渡されたint型の整数が、存在する商品番号かどうか」を確かめるための条件は再検討中です。

(lengthやsizeを超過しているかどうかで確かめる方式だと、現状のテストコードを通過しないため)

wnakayama committed:
FIX productInfoList to productInfoMap

wnakayama committed:
FIX productInfoList to productInfoMap at src

wnakayama commented on Source/VendingMachine/src/pkg/Storage.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.List;
+import java.util.Map;
+
+public class Storage {
+
+    Map<Integer, Product> productInfoMap = new HashMap<Integer, Product>();
+    Map<Integer, Integer> stockMap = new HashMap<Integer, Integer>();
+
+    public void addProductInfo(Integer num, Product product) {
+        productInfoMap.put(num, product);
+    }
+
+    public void chargeStock(Integer num, Integer stock) {
+        stockMap.put(num, stock);
+    }
```
>getProduct 、何番目に追加した商品かを知ってなければ使えないんですが、意図したところかな？

こちらの問題を解消するため、商品情報もListではなくMapで管理するように変更しました。

前回のコミットにあったchargeProductメソッドですが、
「商品情報」と「在庫数」を別々のコレクションで管理しているにもかかわらず
このようなメソッド名では紛らわしいと判断し、「addProductInfo」に変更しました。

yutsunom replied:
Storage クラスでしょ？
ジュース１本ずつ追加するわけだから、charge するのはジュースでは。
１０本増やしました、という場合、（public メソッドにて）わざわざストレージのカウンタをアップしますか？
不整合の元だと思います

wnakayama commented on Source/VendingMachine/src/pkg/Storage.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    Product getProduct(int num) throws WrongProductNumberException {
+        Product pr = productList.get(num - 1);
```
すみません。ここは次回のコミットまでに修正します。
(Listで管理していた時のままでした)
【誤】
  Product pr = productList.get(num - 1);
【正】
 Product pr = productMap.get(num);

- [x] 修正

wnakayama commented on test/pkg/VendingMachineTest.java:
```diff
@@ -23,75 +22,128 @@ public void getDeposit() {
     @Test
     /**
-     * get在庫数のテスト storageのstockMapから正しく在庫数を取得できるか確認する
+     * get在庫数のテスト storageのstockMapから正しく在庫数を取得できるか確認する @throws
```
- [x] 余分にコピー&ペーストされている@throwsを削除

W-Nakayama-24 commented:
VendingMachine, Storage, Checker すべてのテストクラスで
自前のテスト用Product、Stockを用意し、アサーションで確認するテストコードに修正しました。

ハイライトされているコードが必要以上に多くなっていて恐縮ですが、
それぞれのテストでアサーションの直前にProductや、Stockを用意している点をご確認いただければと思います。

W-Nakayama-24 commented:
[11/26 16:30 口頭で頂いたレビューのメモ]
入れ物としてのstorageから、ジュースを出し入れするとき
★現実では1本ずつ出したり入れたりしているはず
今のコードだと、複数本を一気に入れているようで現実に即していない

そうすると何が良いのか?
取り出すときも、1コずつ減らす

何をpublicにするか、何をprivateにするか、要検討

少なくともchargeStockがpublicであることは違和感がある
chargeStockの第二引数はProductであるべき
(ここに、増やしたい数量(stock)を渡すだけでは現実と乖離がある)
- [ ] Productを渡すなら、addProductInfo, chargeStockともに第一引数(Num)は要らなくなるはず
↑getNumが使えるから

- [ ] (Num⇒ ProductID とかのほうが適切では)

- [ ] 複数本まとめて追加するにしても、arrayを使用して、for文を回して1本ずつ入れている状態を実現する

times

iwanam commented:
あれ、図は↓のプルリクで作っているんじゃないの？　こちらにもファイルがあがってるけど何ででしょ？
https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/8

wnakayama committed:
FIX description of throws, typo

wnakayama committed:
FIX getProduct_method

wnakayama committed:
FIX getProduct_method

W-Nakayama-24 commented:
[11/27 11:30 口頭で頂いたレビューのメモ]
今回の自販機システムでは、1スプリントで機能が完結していない
補充するが無いと購入するも本来できないはず

1本ずつ入れることは実現できないとしても、
在庫に溜めておくのはあくまでproductそのものであることを意識して、
せめて在庫数という数字だけが増減している感じは解消したほうが良かった

①StockMapで商品番号と紐づいた在庫数しか管理しないのなら、Storageという名のクラスに、商品情報は不要ではないか

②商品情報もStorageクラスに残すなら、
在庫数管理も数字ではなく商品(Product)自体を出し入れする方がオブジェクト指向と合致している

引数に、Product productではなくArrayList<Product> productsを渡すと、複数渡せるはず

設計の段階で①②のどちらかを実現するべき

W-Nakayama-24 commented:
午前中に口頭でお話ししましたが、現在残っている改善すべき点(https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/9#issuecomment-558500645) は、
一度レビューを通過した設計部分に戻って修正が必要な内容であるため
「設計通りの内容を所定の期間内に実現すること」を優先し、今回は見送らせていただく形をとりたいと思います。
貴重な指摘の数々が反映しきれず、申し訳ありません。

次回以降、重点的に意識して設計を実施します。

W-Nakayama-24 commented:
こちらが、11/26分で発覚した「javadocコメントの過不足」と「引数のタイプミス」に対応した最新のテストコードです。

https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/9/files
